Add upgrade troubleshooting (#172)
harvester/docs: docs/upgrade/troubleshooting.md

## Diagnose the upgrade flow 

A Harvester upgrade process contains several phases. 
    ![](./assets/ts_upgrade_phases.png)

### Phase 1: Provision upgrade repository VM.

The Harvester controller downloads a Harvester release ISO file and uses it to provision a VM. During this phase you can see the upgrade status windows show:

![](./assets/ts_status_phase1.png){: style="width:60%"}

The time to complete the phase depends on the user's network speed and cluster resource utilization. We see failures in this phase due to network speed. If this happens, the user can [start over the upgrade](#start-over-an-upgrade) again.

We can also check the repository VM (named with the format `upgrade-repo-hvst-xxxx`) status and its corresponding pod:

```
$ kubectl get vm -n harvester-system
NAME                              AGE    STATUS     READY
upgrade-repo-hvst-upgrade-9gmg2   101s   Starting   False

$ kubectl get pods -n harvester-system | grep upgrade-repo-hvst
virt-launcher-upgrade-repo-hvst-upgrade-9gmg2-4mnmq     1/1     Running     0          4m44s
```

### Phase 2: Preload container images

> The Harvester controller creates jobs on each Harvester node to download images from the repository VM and preload them. These are container images required for the next release.

Author: Thanks, added all suggestions to a commit.

During this stage you can see the upgrade status windows shows:

![](./assets/ts_status_phase2.png){: style="width:60%"}

> It will take a while for all nodes to preload images. If the upgrade fails at this phase, the user can check job logs in `cattle-system` namespace:

Reviewer: ```suggestion
It will take a while for all nodes to preload images. If the upgrade fails at this phase, the user can check job logs in the `cattle-system` namespace:
```

...
```

### Phase 4: Upgrade nodes

![](./assets/ts_status_phase4.png){: style="width:60%"}

The Harvester controller creates jobs on each node (one by one) to upgrade nodes' OSes and RKE2 runtime. For multi-node clusters, there are two kinds of jobs to update a node:

- **pre-drain** job: live-migrate or shutdown VMs on a node. When the job completes, the embedded Rancher service upgrades RKE2 runtime on a node.
- **post-drain** job: upgrade OS and reboot.

For single-node clusters, there is only one `single-node-upgrade` type job for each node (named with the format `hvst-upgrade-xxx-single-node-upgrade-<hostname>`).

The user can check node jobs by:

```
$ kubectl get jobs -n harvester-system -l harvesterhci.io/upgradeComponent=node
NAME                                  COMPLETIONS   DURATION   AGE
hvst-upgrade-9gmg2-post-drain-node1   1/1           118s       6m34s
hvst-upgrade-9gmg2-post-drain-node2   0/1           9s         9s
hvst-upgrade-9gmg2-pre-drain-node1    1/1           3s         8m14s
hvst-upgrade-9gmg2-pre-drain-node2    1/1           7s         85s

$ kubectl logs -n harvester-system jobs/hvst-upgrade-9gmg2-post-drain-node2
...
```

!!! warning

>     Please do not start over an upgrade if the upgrade fails at this phase.

Reviewer 2: Is there any method to troubleshot during this phase?

Author: This involves Rancher service. We can improve it gradually.

### Phase 5: Clean-up

> The Harvester controller deletes the upgrade repository VM and all files that are not needed anymore.

Reviewer: ```suggestion
The Harvester controller deletes the upgrade repository VM and all files that are no longer needed.
```

## Common operations

### Start over an upgrade

- Log in to a control plane node.

> - List `Upgrade` CRs in the cluster:

Reviewer 2: Maybe display as ordered list is better?

Author: Fixed
